### ISSUE TITLE
extracting reads from (multiple) regions

### ISSUE THREAD

**Issue by dcopetti** [whatshap/whatshap#262]
Hello,

I would like to tag (and then extract separately) long reads based on the haplotypes that I built with longranger wgs for about 900 regions in my assembly. I can't figure out a way to specify the coordinates: if I do it as `OF1p3c_SSc_Bd_SSc_all_1:320000-520000` like the documentation says, I get `ValueError: invalid literal for int() with base 10: '320000-520000'`. If I remove the coordinates and leave only sequence name
`whatshap haplotag --ignore-read-groups --reference reference.fa --regions OF1p3c_SSc_Bd_SSc_all_1 phased_variants.vcf.gz long_reads_aligned.bam`
it works. 

I tried also having `--regions` file with the regions separated by space:
`OF1p3c_SSc_Bd_SSc_all_1:320000-520000 OF1p3c_SSc_Bd_SSc_all_1:770000-970000 OF1p3c_SSc_Bd_SSc_all_1:5570000-5770000`
but it does not accept a file like that How should I encode those intervals?
As a suggestion, I think it would be great if the user could specify the regions in a bed file, for compatibility with other steps.
Thanks,
Dario

**Comment by tobiasmarschall** [comment]
As far as I remember, @ptrebert refactored the region parsing, so I hope he might be able to help.

**Comment by ptrebert** [comment]
It looks like the relevant part of the `haplotag` module has been rewritten again; however, the code looks good to me, and I cannot reproduce the above `ValueError` with the current code base:
```
Region.parse("OF1p3c_SSc_Bd_SSc_all_1:320000-520000 ")
Region("OF1p3c_SSc_Bd_SSc_all_1", 319999, 520000)
```
What version of WhatsHap are you using?

**Comment by dcopetti** [comment]
Hi Peter,
I am using whatshap 1.0.
The command above completed overnight, it printed out this
Found 1 samples in input VCF
Keeping 1 samples for haplo-tagging
Found 0 samples in BAM file
WARNING: Skipping duplicated position 71252637 on chromosome 'OF1p3c_SSc_Bd_SSc_all_1'
Reading alignments and detecting alleles ...
Found 48034 reads covering 62114 variants

== SUMMARY ==
Total alignments processed:                    131049
Alignments that could be tagged:                44757
Alignments spanning multiple phase sets:         1321
haplotag - total processing time: 68.04450273513794
and a bam file where I can not find the phase assignment for each read:
samtools view scf1_test.bam | cut -f1-5,7-9,12-22 | more
8759cdad-4293-46d6-863c-cd11f37c40d8    16      OF1p3c_SSc_Bd_SSc_all_1 169     1       *       0       0       NM:i:275        ms:i:2550       AS:i:2550       nn:i:0  tp:A:P  cm:i:20 s1:i:168        s2:i:185        de:f:0.1034     rl:i:776
43ee15da-8efb-429b-a6a7-b7664cf14f14    16      OF1p3c_SSc_Bd_SSc_all_1 309     39      *       0       0       NM:i:188        ms:i:2134       AS:i:2134       nn:i:0  tp:A:P  cm:i:16 s1:i:150        s2:i:124        de:f:0.0896     SA:Z:OF1p3c_SSc_Bd_SSc_all_93,1866943,+,1627S943M50D4S,1,122;     rl:i:1137
763cbda3-92bb-4a5b-9e3f-3dfd34bde48f    16      OF1p3c_SSc_Bd_SSc_all_1 406     1       *       0       0       NM:i:1750       ms:i:19132      AS:i:19132      nn:i:0  tp:A:P  cm:i:418        s1:i:3502       s2:i:3537       de:f:0.0916     rl:i:4630
d6b64187-097f-4e59-b706-4a198375d42b    16      OF1p3c_SSc_Bd_SSc_all_1 752     2       *       0       0       NM:i:1959       ms:i:16350      AS:i:16350      nn:i:0  tp:A:P  cm:i:325        s1:i:2879       s2:i:2839       de:f:0.1096     rl:i:3499
f8687c7b-59c1-4aee-a2e2-6576cccef455    16      OF1p3c_SSc_Bd_SSc_all_1 1201    1       *       0       0       NM:i:587        ms:i:4624       AS:i:4624       nn:i:0  tp:A:P  cm:i:119        s1:i:965        s2:i:964        de:f:0.1153     rl:i:883
01be917b-0f30-4bf7-bc9c-d751d0d04479    16      OF1p3c_SSc_Bd_SSc_all_1 1336    7       *       0       0       NM:i:415        ms:i:3088       AS:i:3088       nn:i:0  tp:A:P  cm:i:80 s1:i:650        s2:i:642        de:f:0.1213     rl:i:318
c96efe21-df6f-4c9b-89a8-cd071790c446    16      OF1p3c_SSc_Bd_SSc_all_1 1425    13      *       0       0       NM:i:408        ms:i:2644       AS:i:2644       nn:i:0  tp:A:P  cm:i:79 s1:i:645        s2:i:622        de:f:0.1322     rl:i:269
8056b1c0-c20c-4332-81d0-b1e1d66da7c6    2064    OF1p3c_SSc_Bd_SSc_all_1 1570    1       *       0       0       NM:i:115        ms:i:272        AS:i:272        nn:i:0  tp:A:P  cm:i:3  s1:i:41 s2:i:0  de:f:0.1788     SA:Z:OF1p3c_SSc_Bd_SSc_all_93,11856319,-,1039S1281M15D,1,125;OF1p3c_SSc_Bd_SSc_all_12,55116641,-,527M14I1779S,1,61;       rl:i:774
(looks exactly like the input bam file)
So now I would need to extract reads (ONT, the phasing was made with 10x and longranger wgs) from a region, assign to each a phase tag, separate them by phase. Later I will assemble them independently.
I wonder now if this is the best way to proceed or if I should create two pseudoreferences and align reads to both (competitive alignment) to have them separated. But it will take more time of course.
What do you think would be the best way?
Thanks!

**Comment by ptrebert** [comment]
Did you run your test with `--ignore-read-groups` as stated in your first post?
If there is a problem caused by that region name, it may lead to skipping everything. I assume the BAM file is quite large / too large to be shared? Could you align a subset of the reads that fall into the region `OF1p3c_SSc_Bd_SSc_all_1:320000-520000` so we can test what's going on?

**Comment by marcelm** [comment]
@dcopetti Try writing `--region OF1p3c_SSc_Bd_SSc_all_1:320000:520000` (with a colon between start and end position)

WhatsHap 1.0 expects a colon as separator between the start and end coordinate as in `chromosome:1:100`. This was surprising (it is documented that way, though), so I fixed this shortly after 1.0 was released (in commit 2ec693df2cdc3bf16888b6fd0f6ff5fb0b787eb4). In the next version, the spelling `chromosome:1-100` is expected instead. For backwards compatibility, we should actually accept both spellings.

Reading regions from a BED file sounds like a good feature request. I have opened #263 for that.

**Comment by dcopetti** [comment]
Thank you for the feedback!
Here is my homework:
with the ":" between start and stop coords it works:
```
$ whatshap haplotag -o test.bam --ignore-read-groups --reference OF1p3c_SSc_Bd_SSc_all_2925_org_Lngr.fa --regions OF1p3c_SSc_Bd_SSc_all_1:320000:520000 phased_variants.vcf.gz scf1-10.bam
Found 1 samples in input VCF
Keeping 1 samples for haplo-tagging
Found 0 samples in BAM file
Reading alignments and detecting alleles ...
Found 29 reads covering 118 variants

== SUMMARY ==
Total alignments processed:                       184
Alignments that could be tagged:                   29
Alignments spanning multiple phase sets:            0
haplotag - total processing time: 9.171234369277954
```
and the bam file is appropriate.
```
(whatshap_env) bash-4.2$ samtools view test.bam | cut -f1-5,7-9,12-22 | head -n2
f4616ec1-a720-457a-b38f-dfd7bb127ced    0       OF1p3c_SSc_Bd_SSc_all_1 399854  60      *       0       0       NM:i:1230       ms:i:18854      AS:i:18854      nn:i:0  tp:A:P  cm:i:1149       s1:i:7854       s2:i:143        de:f:0.0729     SA:Z:OF1p3c_SSc_Bd_SSc_all_1,4915942,-,12788S4533M225D8S,17,685;  rl:i:1194
9ab69a04-1678-49c6-93bb-98ead913c22f    2048    OF1p3c_SSc_Bd_SSc_all_1 399854  60      *       0       0       NM:i:268        ms:i:2350       AS:i:2350       nn:i:0  tp:A:P  cm:i:86 s1:i:732        s2:i:60 de:f:0.1103     SA:Z:xae_seqs,1087733,-,1953S3732M162D20S,1,750;  rl:i:635
(whatshap_env) bash-4.2$ samtools view test.bam | cut -f1-5,7-9,12-22 | tail -n 2
3a850d95-9078-47df-aa78-019c280baace    0       OF1p3c_SSc_Bd_SSc_all_1 462005  46      *       0       0       NM:i:540        ms:i:2892       AS:i:2892       nn:i:0  tp:A:P  cm:i:62 s1:i:604        s2:i:508        de:f:0.1363     rl:i:459
ca26ddc9-71fc-436f-999e-c7ac3ae49675    16      OF1p3c_SSc_Bd_SSc_all_1 462874  28      *       0       0       NM:i:237        ms:i:3130       AS:i:3130       nn:i:0  tp:A:P  cm:i:82 s1:i:761        s2:i:683        de:f:0.0867     rl:i:795
```
which is OK since the scf has two gaps in the interval I input:
```
OF1p3c_SSc_Bd_SSc_all_1 292117  399853
OF1p3c_SSc_Bd_SSc_all_1 465421  753416
```
However, how can I split reads based on the haplotype they belong to?
I will open a separate question thread for my next steps
